dali/operators/generic/slice/subscript.h

```diff
@@ -0,0 +1,325 @@
+  }
+
+  bool IsDefined() const {
+    return src != ArgSource::None;
+  }
+  string name;
+  ArgSource src = ArgSource::None;
+  vector<int64_t> values;
+  int64_t scalar_value = 0;
+
+  void Load(const ArgumentWorkspace &ws, int batch_size) {
+    if (src == ArgSource::Input) {
+      const TensorVector<CPUBackend> &inp = ws.ArgumentInput(name);
+      const auto &shape = inp.shape();
+
+      int n = batch_size;
+      DALI_ENFORCE(shape.num_samples() == n, make_string(
+          "Unexpected number of samples in argument `", name, "`. Got: ",
+          shape.num_samples(), ", expected ", n, "."));
+
+      DALI_ENFORCE(shape.sample_dim() == 0,
+        make_string("Array indices must be scalar (0D). Got ", shape.sample_dim(), "D tensor."));
+
+      values.resize(n);
+      DALIDataType type_id = inp.type().id();
+      TYPE_SWITCH(type_id, type2id, T, (INTEGER_TYPES), (
+        for (int i = 0; i < n; i++) {
+          // TODO(michalz): Add tensor<T> and mutable_tensor<T> to TensorVector?
+          values[i] = ConvertSat<int64_t>(static_cast<const T*>(inp.raw_tensor(i))[0]);
+        }
+      ), DALI_FAIL(make_string("Array index must be of integral type. Got: ", type_id)));  // NOLINT
+    } else if (src == ArgSource::Value) {
+      values.resize(batch_size, scalar_value);
+    }
+  }
+};
+
+struct SubscriptInfo {
+  void Init(const OpSpec &spec, int i) {
+    at.Init(spec, make_string("at_", i));
+    lo.Init(spec, make_string("lo_", i));
+    hi.Init(spec, make_string("hi_", i));
+    step.Init(spec, make_string("step_", i));
+
+    if (at.IsDefined()) {
+      if (lo.IsDefined() || hi.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified both as an index and as a range."));
+      }
+      if (step.IsDefined()) {
+        DALI_FAIL(make_string("The subscriptg for dimension ", i,
+                              " is specified as an index - it cannot have a step."));
+      }
+    }
+  }
+
+  bool IsDefined() const {
+    return at.IsDefined() || lo.IsDefined() || hi.IsDefined() || step.IsDefined();
+  }
+
+  bool IsIndex() const {
+    return at.IsDefined();
+  }
+
+  bool IsRange() const {
+    return !IsIndex() && IsDefined();
+  }
+
+  SubscriptArg at, lo, hi, step;
+
+  void Load(const ArgumentWorkspace &ws, int batch_size) {
+    at.Load(ws, batch_size);
+    lo.Load(ws, batch_size);
+    hi.Load(ws, batch_size);
+    step.Load(ws, batch_size);
+  }
+};
+
+
+template <typename Backend>
+class TensorSubscript : public Operator<Backend> {
+ public:
+  explicit TensorSubscript(const OpSpec &spec) : Operator<Backend>(spec) {
+    InitArgs();
+  }
+
+  USE_OPERATOR_MEMBERS();
+
+  void InitArgs() {
+    subscripts_.resize(kMaxSubscripts);
+    int num_subscripts = 0;
+    for (int i = 0; i < kMaxSubscripts; i++) {
+      SubscriptInfo &s = subscripts_[i];
+      s.Init(spec_, i);
+
+
+      if (s.IsDefined()) {
+        num_subscripts = i + 1;
```

Comment:
```suggestion
        max_subscripts = i + 1;
```
As it is not a number because some subscripts may be not defined.

Reply:
Mixed feelings here. It's used as the number (see `subscripts_.resize(num_subscripts);`). I would gladly use input ndim here _if I had it_ - but it's not available at construction.